sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -95,13 +96,40 @@ object AnalysisContext {
 class Analyzer(
     catalog: SessionCatalog,
     conf: SQLConf,
-    maxIterations: Int)
+    maxIterations: Int,
+    lookupCatalog: Option[(String) => CatalogPlugin] = None)
   extends RuleExecutor[LogicalPlan] with CheckAnalysis {
 
   def this(catalog: SessionCatalog, conf: SQLConf) = {
     this(catalog, conf, conf.optimizerMaxIterations)
   }
 
+  def this(lookupCatalog: Option[(String) => CatalogPlugin], catalog: SessionCatalog,
+      conf: SQLConf) = {
+    this(catalog, conf, conf.optimizerMaxIterations, lookupCatalog)
+  }
+
+  object CatalogRef {
+    def unapply(parts: Seq[String]): Option[(Option[CatalogPlugin], CatalogIdentifier)] =
```

Comment:
I like this. It provides a good way to add support in match expressions. You might consider adding one that translates to `TableIdentifier` as well:

```scala
object V1Identifier {
  def unapply(parts: Seq[String]): Option[TableIdentifier] = ...
}
```

That would match if there are only 2 identifier parts and the first is not a known catalog name.

Reply:
To clarify because Github isn't putting your comment in a thread, is this in response to https://github.com/apache/spark/pull/23848#discussion_r266179932? If so, is "I like this" meaning "I like the case class" that above comment has suggested, or, that you prefer the way it is currently?

Just trying to disambiguate.

Reply:
This wasn't part of the other thread. I like using an extractor as a helper for writing rules. This will help us produce rules that are cleaner.